File: src/test/java/sk/george/intervierw/ApplicationTest.java
package sk.george.intervierw;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import java.util.List;

import org.junit.jupiter.api.Test;

import sk.george.intervierw.io.input.TestConsoleInputHandler;
import sk.george.intervierw.io.input.TestFileInputHandler;
import sk.george.intervierw.io.output.OutputHandler;
import sk.george.intervierw.io.output.TestOutputHandler;
import sk.george.intervierw.word_counter.WordCounter;

class ApplicationTest {

    @Test
    void run_printsWordsCount_inputFromUser() {
        // Init data
        TestOutputHandler outputHandler = new TestOutputHandler();
        String userInput = "Mary had a little lamb";
        Application application = createApp(userInput, null, outputHandler);

        // Start the application
        application.run(null);
        // Assert the result
        assertNotNull(outputHandler.getMessages());
        assertEquals(2, outputHandler.getMessages().size());
        assertEquals("Number of words: 5", outputHandler.getMessages().get(1));
    }

    private Application createApp(String mockInput, List<String> mockInputs, OutputHandler outputHandler) {
        return new Application(
                new TestConsoleInputHandler(mockInput),
                new TestFileInputHandler(mockInputs),
                outputHandler,
                new WordCounter()
        );
    }
}

File: src/test/kotlin/ApplicationTest.kt
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.io.ByteArrayOutputStream
import java.nio.file.Files
import kotlin.io.path.absolutePathString
import kotlin.io.path.deleteIfExists
import kotlin.io.path.writeText

class ApplicationTest {

    @Test
    fun `count words in sentence`() {
        // Given
        val askForInputStream = ByteArrayOutputStream(1024)
        val resultStream = ByteArrayOutputStream(1024)

        askForInputStream.bufferedWriter().use { askForInputStreamWriter ->
            resultStream.bufferedWriter().use { resultStreamWriter ->
                val counter = Application(
                    inputTextReader = InputTextReaderImpl(
                        arguments = emptyArray(),
                        textInputStream = "Mary had a little lamb".byteInputStream(),
                        writer = askForInputStreamWriter,
                    ),
                    stopWordsReader = ClasspathStopWordsReader(),
                    resultWriter = StreamResultWriter(
                        writer = resultStreamWriter
                    ),
                )

                // When
                counter.run()
            }
        }

        // Then
        Assertions.assertEquals("Enter text: ", askForInputStream.toString())
        Assertions.assertEquals("Number of words: 4, unique: 4\n", resultStream.toString())
    }

    @Test
    fun `count words from provided file`() {
        // Given
        val askForInputStream = ByteArrayOutputStream(1024)
        val resultStream = ByteArrayOutputStream(1024)

        val testInputFile = Files.createTempFile("testInputFile", "txt")
        testInputFile.writeText(
            """
            Mary had
            a little
            lamb
        """.trimIndent()
        )

        askForInputStream.bufferedWriter().use { askForInputStreamWriter ->
            resultStream.bufferedWriter().use { resultStreamWriter ->
                val counter = Application(
                    inputTextReader = InputTextReaderImpl(
                        arguments = arrayOf(testInputFile.absolutePathString()),
                        textInputStream = "Mary had a little lamb".byteInputStream(),
                        writer = askForInputStreamWriter
                    ),
                    stopWordsReader = ClasspathStopWordsReader(),
                    resultWriter = StreamResultWriter(
                        writer = resultStreamWriter
                    ),
                )
                // When
                counter.run()
            }
        }

        // Then
        Assertions.assertEquals("", askForInputStream.toString())
        Assertions.assertEquals("Number of words: 4, unique: 4\n", resultStream.toString())
        testInputFile.deleteIfExists()
    }

}